testing/pike: new aport

grodzik opened the pull request:
https://pike.lysator.liu.se
Pike Programing language

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,235 @@
+pkgdesc="Pike Programing language"
+url="https://pike.lysator.liu.se"
+arch="x86_64"
+license="MIT"
```
The https://pike.lysator.liu.se/download/ site doesn't mention MIT license.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,235 @@
+unpack() {
+	cd "$builddir"
+	tar xzf Pike-v${pkgver}.tar.gz --strip-components=1
+}
```
You'll want to use a `prepare` function instead.

```
prepare() {
  default_prepare

  cd "$builddir"
  : # your other custom commands
}
```

tcely replied:
You shouldn't need this `unpack` function at all.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,235 @@
+			$pkgname-doc \
+			$pkgname-core"
+source="http://pike.lysator.liu.se/pub/pike/all/${pkgver}/Pike-v${pkgver}.tar.gz"
+builddir="$srcdir/"
```
`builddir="$srcdir/Pike-v${pkgver}"`

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,235 @@
+
+package() {
+	cd "$builddir"
+	depends=""
```
Do you have a reason to reset `depends` from the top of the file here? I'd be confused if I added an item to `depends` as I normally would at the top and it didn't end up in the built package.

grodzik replied:
Probably some kind of leftover from development

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,235 @@
+_install() {
+	local bn=$(basename "$1")
+	local dn=$(dirname "$1")
+	for item in `find $dn -maxdepth 1 -name "$bn"`; do
```
This is not a particularly safe construction. Consider using something like this:

```
find "$dn" -maxdepth 1 -name "$bn" | while IFS= read item; do _move_item "$item" "$2"; done
```

grodzik replied:
I'll fix that, but can you share why this isn't safe?

grodzik replied:
I can change that, no problem. However I would be grateful for explanation why `for` approach is unsafe?

tcely replied:
The path in `$dn` is not under your control and neither is the `find` output.

The easiest problem to understand with that `for` construction is word splitting. You're depending on splitting on `\n` for each line `find` prints, but word splitting also happens on any embedded spaces too.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,235 @@
+	cd "$pkgdir"
+	pkgdesc="$pkgname image handling modules"
+	depends="pike-core=$pkgver-r$pkgrel giflib-utils libjpeg tiff libwebp"
+	_install "usr/lib/pike/?.?/modules/Image.pmod" "$subpkgdir"
```
With the quotes, the first argument to `_install` is likely not doing what you expect.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,235 @@
+	_install "usr/lib/pike/include/*.h" "$subpkgdir"
+	_install "usr/include/pike/*.h" "$subpkgdir"
+	mv "$subpkgdir/usr/lib/pike/include/regexp.h" "usr/lib/pike/include/"
+}
```
You'd probably be better off letting the `default_dev` function do its thing.

grodzik replied:
I actually need the `regexp.h` file inside `pike-core` package. Or will
```
dev() {
    default_dev
    mv .....
}
```
do the trick?

tcely replied:
Yep. That'll get it done.

https://github.com/alpinelinux/abuild/blob/c0a862930c1bd033be30f2c48e91a6f7e76a5a58/abuild.in#L1706

tcely replied:
If you want `pike-core` to split before `pike-dev` just change the order in the `subpackages` list.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,235 @@
+	gzip usr/share/man/man1/pike.1
+	mv usr/doc/ "$subpkgdir/usr/share/"
+	mv usr/share/man/ "$subpkgdir/usr/share/"
+}
```
Letting the `default_doc` function handle this would be better.

tcely replied:
The only thing I see that might not be handled by `default_doc` is the `usr/doc` directory. You could change the location of `usr/doc` in the `package` function to `usr/share/doc` so that it is split without any need for your own `doc` implementation.

grodzik commented:
As for number of packages, Pike has a lot of dependencies, taking a lot of data, not necessarily needed to have all the time. I don't even made all the packages available, I'll try to add them later. In general Pike is quite a modular package, thus it would be good to use that well. 

My intention with `pike` package was to act only as a meta-package, instead of creating `pike-full` package, and `pike-core` for the very basic stuff.

tcely commented:
I still think you're making too many complications for yourself by splitting out so much right away. Having `pike` with most of the files, `pike-dev` and `pike-doc` as `subpackages` would be a good starting point.

grodzik commented:
I've fixed issues you've raised. Thanks for reviewing. I removed `-dev` package, since some header files where still needed and that was creating a bit of mess. I removed subpackages that were not depending on any additional libraries, like `-fuse`.  And I've fixed creating documentation (thanks to Gentoo ebuild)

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+	local dn=$(dirname "$1")
+	local dst="$2/$dn"
+	mkdir -p "$dst"
+	mv "$1" "$dst/$bn"
```
`mv "$1" "$dst/$3"` would let you optionally rename the destination file.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+	cd "$pkgdir"
+	pkgdesc="$pkgname image handling modules"
+	depends="pike-core=$pkgver-r$pkgrel giflib-utils libjpeg tiff libwebp"
+	_install "usr/lib/pike/?.?/modules/Image.pmod" "$subpkgdir"
```
Does this work the way you are expecting? I don't think the quoted `?.?` will match.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+bzip2() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname bzip2 support"
+	depends="pike-core=$pkgver-r$pkgrel bzip2"
```
Unless you need the bzip2 binaries, you should remove `bzip2` from depends. `libbz2` is being depended on automatically.

> \>>> pike-bzip2*: Tracing dependencies...
	pike-core=8.0.610-r0
	bzip2
	so:libbz2.so.1

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+sqlite() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname SQLite support"
+	depends="pike-core=$pkgver-r$pkgrel sqlite-libs"
```
`sqlite-libs` is pulled in automatically. There is no need to list it in depends.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+svg() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname SVG support"
+	depends="pike-core=$pkgver-r$pkgrel librsvg"
```
`librsvg` is not needed here.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+	pkgdesc="Pike's core files and executables"
+	depends="gmp nettle nettle-utils zlib readline libgcrypt"
+	mkdir -p "$subpkgdir"
+	mv usr "$subpkgdir/"
```
I'm going to suggest leaving `usr/bin` in `pike` package.

```
mkdir -p "$subpkgdir/usr"
mv usr/include usr/lib "$subpkgdir/usr/"
```

grodzik replied:
Ok, so maybe get rid of `pike-core` and leave core stuff in `pike`, and add something like `pike-full` meta package that will pull in all additional `pike-` packages?

tcely replied:
I'd suggest `pike-modules` so that you could choose to install `pike` and specific individual modules or `pike` and `pike-modules` which would depend on all the split up modules.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+core() {
+	cd "$pkgdir"
+	pkgdesc="Pike's core files and executables"
+	depends="gmp nettle nettle-utils zlib readline libgcrypt"
```
`gmp`, `zlib` are already included automatically.

I don't see an immediate use for `libgcrypt`. Would you explain why it's here?

jirutka commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+			sqlite-libs \
+			readline \
+			zlib-dev \
+			librsvg-dev"
```
It'd not needed to escape newlines inside a string. Please remove these backslahes and use only one tab for indentation.

jirutka commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,200 @@
+
+package() {
+	cd "$builddir"
+	for pkg in $(printf "%s" "$subpackages" | tr ' ' '\n'); do
```
`pkg` should be declared as local (`local pkg; for pkg in ...`).

tcely commented on community/pike/APKBUILD:
```diff
@@ -89,96 +85,89 @@ build() {
-	for pkg in $(printf "%s" "$subpackages" | tr ' ' '\n'); do
-		depends="$depends $pkg=$pkgver-r$pkgrel"
-	done
+	depends="gmp nettle nettle-utils zlib readline"
```
Please put the `depends` for the package in the `depends` variable at the top of the file. This information is used to control the order in which builds happen from what I understand.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+url="https://pike.lysator.liu.se"
+arch="x86_64"
+license="GPL-2.0-or-later"
+options="!check" # there is no tests for those packages
```
grammar: there are no tests for this package

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+# Maintainer: Paweł Tomak <pawel@tomak.eu>
+pkgname=pike
```
`_pkgname=Pike`

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+install=
+subpackages="$pkgname-doc $pkgname-full"
+source="http://pike.lysator.liu.se/pub/pike/all/${pkgver}/Pike-v${pkgver}.tar.gz"
+builddir="$srcdir/Pike-v$pkgver"
```
```
source="http://pike.lysator.liu.se/pub/${pkgname}/all/${pkgver}/${_pkgname}-v${pkgver}.tar.gz"
builddir="${srcdir}/${_pkgname}-v${pkgver}"
```

jirutka replied:
Please don't promote unnecessary curly braces around variables, respect our code-style.

tcely replied:
I still don't get why you insist on everyone keeping these variable parsing rules in our heads, but I'll edit it to respect the style as you requested.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+arch="x86_64"
+license="GPL-2.0-or-later"
+options="!check" # there is no tests for those packages
+depends=
```
depends from `package` function should go here

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+do
+	subpackages="$subpackages $pkgname-${_name}"
+	_full_deps="${_full_deps} $pkgname-${_name}=$pkgver-r$pkgrel"
+done
```
Consider adding `$pkgname-dev` to the end of `subpackages` to split out the include files

grodzik replied:
There are at least two headers, that would have to be moved back manually to main package. Pike's preprocesor needs them while parsing some core files.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+		--without-freetype \
+		--without-vcdiff \
+		--with-_Image_WebP \
+		--with-zlib"
```
Do you need to escape the new lines in this quoted string?

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+		--without-vcdiff \
+		--with-_Image_WebP \
+		--with-zlib"
+	PATH="${builddir}/bin:${PATH}" make doc || die "doc failed"
```
When make doc returns an error that error having come from make doc should be fairly self evident.

What value does `|| die "doc failed"` add here?

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+bzip2() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname support"
+	depends="pike=$pkgver-r$pkgrel bzip2"
```
You don't need this `depends` line here at all. `bzip2` is the CLI tool for the `libbz2` library that is being found and depended on automatically.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+gdbm() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname GDBM support"
+	depends="pike=$pkgver-r$pkgrel gdbm"
```
Compare how `depends="pike"` works here in your next build.

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+sdl() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname SDL support"
+	depends="pike-image=$pkgver-r$pkgrel sdl sdl_mixer"
```
`depends="pike"` again

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,187 @@
+pcre() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname PCRE support"
+	depends="pike=$pkgver-r$pkgrel pcre"
```
This is pulled in by the `so:libpcre.so.1` that is added automatically, so listing `pcre` is redundant.

`depends="pike"`

grodzik commented:
Hi, is there anything else to fix here?

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,188 @@
+
+_install() {
+	local _arg
+	for _arg in "$@"; do
```
The `in "$@"` isn't needed, but won't hurt anything.

Try this and see for yourself.
`$ /bin/sh -c 'foo() { local arg; for arg; do echo $arg; done; }; foo a b c;'`

jirutka replied:
I know that this shortened syntax is defined in POSIX, I just don't find it intuitive and reasonable. That's just my opinion.

tcely commented:
@grodzik Is it possible to clean these up?

```
Could not resolve Getopt. (Perhaps the installed pike tree has been moved.)
lib/master.pike:3163: master()->main_resolv("Getopt",UNDEFINED)
Dumping of some modules failed (not fatal) (0x0000000a):
  "/usr/lib/pike/modules/Crypto.pmod/Random.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/SHA3_512.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/SHA256.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/DH.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/PGP.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/Substitution.pike"
  "/usr/lib/pike/modules/Crypto.pmod/ChaCha20.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/SHA3_384.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/AES.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/SHA3_224.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/SHA512.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/SALSA20R12.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/ECC.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/SALSA20.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/module.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/Koremutake.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/Camellia.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/DSA.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/Arcfour.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/DES3.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/MD2.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/Twofish.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/SHA224.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/CAST.pmod"
  "/usr/lib/pike/modules/Crypto.pmod/GOST94.pmod"
```

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,188 @@
+	zlib-dev
+	librsvg-dev"
+install=
+subpackages="$pkgname-doc $pkgname-full"
```
The build log is suggesting `$pkgname-full::noarch`

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,188 @@
+gdbm() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname GDBM support"
+	depends="pike gdbm"
```
`gdbm` is not needed in depends here. See https://pkgs.alpinelinux.org/contents?file=libgdbm.so.4&path=&name=&branch=edge&arch=x86_64

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,188 @@
+gl() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname OpenGL/GLUT support"
+	depends="pike freeglut mesa-gl libsm"
```
`mesa-gl` is not needed in this depends

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,188 @@
+sdl() {
+	cd "$pkgdir"
+	pkgdesc="$pkgname SDL support"
+	depends="pike-image sdl sdl_mixer"
```
`sdl` and `sdl_mixer` aren't needed in this depends line

tcely commented on community/pike/APKBUILD:
```diff
@@ -0,0 +1,188 @@
+arch="x86_64"
+license="GPL-2.0-or-later"
+options="!check" # there are no tests for those packages
+depends="gmp nettle nettle-utils readline"
```
`gmp` (`so:libgmp.so.10`) and `nettle` (`so:libnettle.so.6`) can be removed from this `depends` list

grodzik commented:
Thanks for the review @tcely and @jirutka. Anything else needs to be done from my side to push this further?

ncopa commented:
It would be nice to merge all the commits. We should also move it to `testing/pike` instead of `community/pike`, since it is a new package. I suppose I can do that for you.

We can move it to community once it has been verified that it works, even when built with the official builders.

Thanks!

ncopa commented:
fixed with eec7db60bd8274db29ce5731c26019778a22f28e
